Adding Types for siren-parser

dillonredding opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

peterblazejewicz commented on types/siren-parser/superagent.d.ts:
```diff
@@ -0,0 +1,6 @@
+
+export function parse(res: string): Entity;
+export function parse(res: Response, fn: (err: Error | null, body: Entity) => void): void;
+export function perform(request: any, action: Action): SuperAgentRequest;
```
It looks the SuperAgentRequest needs to be augmented here to have the `submit` method:
https://github.com/Brightspace/node-siren-parser/blob/master/src/superagent.js#L37
Otherwise sample code from Readme won't work:
https://github.com/Brightspace/node-siren-parser/blob/master/README.md#L88

dillonredding replied:
I did notice that while writing the test code, but the `submitHelper` function isn't exported and it's not referenced inside that file, so I'm not sure how that's running. Could that be a "bug" (in the docs)?

peterblazejewicz commented on types/siren-parser/index.d.ts:
```diff
@@ -0,0 +1,143 @@
+// Project: https://github.com/Brightspace/node-siren-parser
+// Definitions by: Dillon Redding <https://github.com/dillonredding>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// Minimum TypeScript Version: 3.7
```
```suggestion
// TypeScript Version: 3.0
```
this should work ok, as the `superagent` states this version, this seems to be only package-depending, not depending on the TS language features, correct?

dillonredding replied:
I added that per a recommendation from the linter:

```
$ npm run lint siren-parser

> definitely-typed@0.0.3 lint C:\dev\DefinitelyTyped
> dtslint types "siren-parser"

Error: C:/dev/DefinitelyTyped/types/siren-parser/siren-parser-tests.ts:103:6     
ERROR: 103:6   expect  Compile error in typescript@3.6 but not in typescript@3.7.
Fix with a comment '// Minimum TypeScript Version: 3.7' just under the header.   
Expression expected.
ERROR: 103:10  expect  Compile error in typescript@3.6 but not in typescript@3.7.
Fix with a comment '// Minimum TypeScript Version: 3.7' just under the header.   
':' expected.
...
```

Seems to be where I'm doing `variable?.prop;`. To be honest, I'm just getting back into TypeScript after a few years. Is there a better approach here?

peterblazejewicz replied:
I see, but the problem is only in the tests (Elvis operator). I usually opt to older syntax in test, to not limit the scope of the usage of the type definition itself.

dillonredding replied:
Gotcha. I replaced the `?.`s with `if` checks and lowered the minimum version.

peterblazejewicz commented:
So it works OK in ES6 like syntax, but the package is supposed to support UMD-ish scenarios (as per README). As this is OK and you're the owner, can I only suggest to try to rewrite the PR to support all three scenarios (ES5/global/CMJ/ES6):
I would start with `index.d.ts` rewritten to:
```ts
// Type definitions for siren-parser 8.2
// Project: https://github.com/Brightspace/node-siren-parser
// Definitions by: Dillon Redding <https://github.com/dillonredding>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
// TypeScript Version: 3.0

import Action from './Action';
import Link from './Link';

declare global {
    namespace D2L {
        namespace Hypermedia {
            namespace Siren {
                function Parse(siren: string | object): Entity;
            }
        }
    }
}

export default function SirenParse(siren: string | object): Entity;

export class Entity {
    constructor(siren: string | object);
    class?: string[];
    properties?: object;
    entities?: Entity[];
    links?: Link[];
    actions?: Action[];
    title?: string;
    // embedded link properties
    rel?: string[];
    href?: string;
    type?: string;

    hasAction(actionName: string | RegExp): boolean;
    hasActionByName(actionName: string | RegExp): boolean;
    hasActionByClass(actionClass: string | RegExp): boolean;
    hasActionByMethod(actionMethod: string | RegExp): boolean;
    hasActionByType(actionType: string | RegExp): boolean;
    hasClass(cls: string | RegExp): boolean;
    hasEntity(entityRel: string | RegExp): boolean;
    hasEntityByRel(entityRel: string | RegExp): boolean;
    hasSubEntityByRel(entityRel: string | RegExp): boolean;
    hasEntityByClass(entityClass: string | RegExp): boolean;
    hasSubEntityByClass(entityClass: string | RegExp): boolean;
    hasEntityByType(entityType: string | RegExp): boolean;
    hasSubEntityByType(entityType: string | RegExp): boolean;
    hasLink(linkRel: string | RegExp): boolean;
    hasLinkByRel(linkRel: string | RegExp): boolean;
    hasLinkByClass(linkClass: string | RegExp): boolean;
    hasLinkByType(linkType: string | RegExp): boolean;
    hasProperty(property: string | RegExp): boolean;
    getAction(actionName: string | RegExp): Action | undefined;
    getActionByName(actionName: string | RegExp): Action | undefined;
    getActionByClass(actionClass: string | RegExp): Action | undefined;
    getActionsByClass(actionClass: string | RegExp): Action[];
    getActionByClasses(actionClasses: ReadonlyArray<string | RegExp>): Action | undefined;
    getActionsByClasses(actionClasses: ReadonlyArray<string | RegExp>): Action[];
    getActionByMethod(actionMethod: string | RegExp): Action | undefined;
    getActionsByMethod(actionMethod: string | RegExp): Action[];
    getActionByType(actionType: string | RegExp): Action | undefined;
    getActionsByType(actionType: string | RegExp): Action[];
    getLink(linkRel: string | RegExp): Link | undefined;
    getLinks(linkRel: string | RegExp): Link[];
    getLinkByRel(linkRel: string | RegExp): Link | undefined;
    getLinksByRel(linkRel: string | RegExp): Link[];
    getLinkByRels(linkRels: ReadonlyArray<string | RegExp>): Link | undefined;
    getLinksByRels(linkRels: ReadonlyArray<string | RegExp>): Link[];
    getLinkByClass(linkClass: string | RegExp): Link | undefined;
    getLinksByClass(linkClass: string | RegExp): Link[];
    getLinkByClasses(linkClasses: ReadonlyArray<string | RegExp>): Link | undefined;
    getLinksByClasses(linkClasses: ReadonlyArray<string | RegExp>): Link[];
    getLinkByType(linkType: string | RegExp): Link | undefined;
    getLinksByType(linkType: string | RegExp): Link[];
    getSubEntity(entityRel: string | RegExp): Entity | undefined;
    getSubEntities(entityRel: string | RegExp): Entity[];
    getSubEntityByRel(entityRel: string | RegExp): Entity | undefined;
    getSubEntitiesByRel(entityRel: string | RegExp): Entity[];
    getSubEntityByRels(entityRels: ReadonlyArray<string | RegExp>): Entity | undefined;
    getSubEntitiesByRels(entityRels: ReadonlyArray<string | RegExp>): Entity[];
    getSubEntityByClass(entityClass: string | RegExp): Entity | undefined;
    getSubEntitiesByClass(entityClass: string | RegExp): Entity[];
    getSubEntityByClasses(entityClasses: ReadonlyArray<string | RegExp>): Entity | undefined;
    getSubEntitiesByClasses(entityClasses: ReadonlyArray<string | RegExp>): Entity[];
    getSubEntityByType(entityType: string | RegExp): Entity | undefined;
    getSubEntitiesByType(entityType: string | RegExp): Entity[];
}
```
and all other modules moved to related files (`Action.d.ts`, `Field.d.ts`, `Link.d.ts`, 'superagent.d.ts`), as this would reflect the package layout and shoudl work OK wiht copy pasted imports from native package?:
```ts
import Action from './Action';
import Link from './Link';
```
```ts
import Field from './Field';
export default interface Action {
```
etc.
What do you think? As I said, you're the owner/author, so current defnition si also OK.

dillonredding commented:
@peterblazejewicz, I'm totally fine with breaking it out like that. I wasn't 100% sure if I should since the `Action`, `Field`, etc. modules didn't appear to be intended for _direct_ use, at least based on the README. I'm very new to type declarations, so I'll defer to your recommendation on this :)

Also, I wasn't totally sure if this _actually_ fit the UMD pattern because there's [an `import` in `global.js`](https://github.com/Brightspace/node-siren-parser/blob/a5adfb8d659be865bba662118ec35804e8b30c7b/global.js#L1), but I just saw that it's using `deumdify`, so I guess it qualifies.

Lastly, do you have any recommendations for the [`chaiPlugin` module](https://github.com/Brightspace/node-siren-parser/blob/a5adfb8d659be865bba662118ec35804e8b30c7b/src/chaiPlugin.js)? Being new to type declarations, I wasn't sure how this should work since I'm not familiar with chai, and it seems to require extending other type declarations.

peterblazejewicz commented:
> Lastly, do you have any recommendations for the [`chaiPlugin` module](https://github.com/Brightspace/node-siren-parser/blob/a5adfb8d659be865bba662118ec35804e8b30c7b/src/chaiPlugin.js)? Being new to type declarations, I wasn't sure how this should work since I'm not familiar with chai, and it seems to require extending other type declarations.

This could come later I guess. 
Probably - as I've never augmented/extended assertion library yet, you would have to (in this additional module):
- augment Chai methods to accept and return types from `siren-parser`.
- sample can be seen in the `@types/sinon-chai`, where `chai` is agumented and `sinon-chai` is exported (same should be done with `siren-parser` I believe:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/types/sinon-chai/sinon-chai.d.ts

dillonredding commented:
Hey, @peterblazejewicz. I split out the modules as you suggested and added re-exports so users can simply export everything from `siren-parser` directly (not sure if that's recommended, but seems like a good idea). I also added support for the global approach, and I _believe_ I got the chai plugin types working correctly (no guarantees though :).
